### PR TITLE
Generate native ctor only when a ctor with required signature exists in base types

### LINKED ISSUE #2937: NativeCtorsGenerator generates invalid constructor for Android
## Current behavior

I have the following type inside a Uno Platform class library.

```
public partial class GameBoardCanvas : SkiaSharp.Views.Android.SKGLTextureView
{
   public GameBoardCanvas(Context context) 
       : base(context)
   {
   }
 
   public GameBoardCanvas(Context context, Android.Util.IAttributeSet attributeSet) 
       : base(context, attributeSet)
   {
   }
} 

```

Unfortunately, when I attempt to compile it, I get errors from `NativeCtorsGenerator`:

> error CS1503: Argument 1: cannot convert from 'System.IntPtr' to 'Android.Content.Context'
> error CS1503: Argument 2: cannot convert from 'Android.Runtime.JniHandleOwnership' to 'Android.Util.IAttributeSet'

This is due to the fact that `SKGLTextureView` does not have the `Context, JniHandleOwnership` constructor.

## Expected behavior

`NativeCtorsGenerator` should skip the unsupported constructor.

## How to reproduce it (as minimally and precisely as possible)

Add the class above to a Uno Platform project (and reference `SkiaSharp` and `SkiaSharp.Views`). Then compile.

## Environment

Nuget Package: 2.2.dev

Package Version(s): 2.2.dev

Affected platform(s):

- [ ] iOS
- [x] Android
- [ ] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #2937

## PR Type

What kind of change does this PR introduce?

- Bugfix
- Refactoring

## What is the current behavior?

- Source generator generates the following even if the base type doesn't contain a ctor with the needed signature, resulting in a compile-time error:

    ```csharp
    public ClassName(IntPtr javaReference, global::Android.Runtime.JniHandleOwnership transfer) : base (javaReference, transfer) { }
    ```

## What is the new behavior?

- Source generator doesn't add any sources for this case.
- Refactoring:
    - `Is` method implementation and name didn't match documentation comment. I renamed it to `InheritsFrom` to match the doc comment and modified the implementation for that too. This needs a careful review that all calls to the method are good with the modified behavior.
    - Removed unused field `_panelSymbol`

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Fixes #2937

Internal Issue (If applicable):

### CODE REVIEW

**Comment by Youssef1313** [comment]
@jeromelaban Can you help with writing tests here? Thanks!

File: src/SourceGenerators/Uno.UI.SourceGenerators.Tests/Uno.UI.SourceGenerators.Tests.csproj [hunk 1]
```diff
@@ -11,7 +11,10 @@
 		<PackageReference Include="MSTest.TestAdapter" Version="2.2.5" />
 		<PackageReference Include="MSTest.TestFramework" Version="2.1.2" />
 		<PackageReference Include="coverlet.collector" Version="3.1.0" />
-		<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.6" />
+		<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.8.0" />
+		<PackageReference Include="Microsoft.CodeAnalysis.CSharp.SourceGenerators.Testing.MSTest" Condition="'$(TargetFramework)'=='net5.0'" Version="1.1.0" />
+		<!-- To have access to Mono.Android.dll from tests -->
+		<!--<Reference Include="$(MSBuildExtensionsPath)\..\Common7\IDE\ReferenceAssemblies\Microsoft\Framework\MonoAndroid\v10.0\Mono.Android.dll" />-->
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Tests/Uno.UI.SourceGenerators.Tests.csproj:17]: What I was trying to do here is to do something like the following:

```csharp
			await new Verify.Test
			{
				TestState =
				{
					Sources = { @"
using Android.Views;

namespace MyApp
{
	internal partial class GameBoardCanvas : TextureView
	{
	}
}
" },
					GeneratedSources =
					{
					},
					AdditionalReferences =
					{
						MetadataReference.CreateFromFile(typeof(Android.Views.TextureView).Assembly.Location),
					},
				},
			}.RunAsync();
```

File: src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/SymbolExtensions.cs [hunk 2]
```diff
@@ -91,7 +91,7 @@ public static IEnumerable<IEventSymbol> GetEvents(INamedTypeSymbol? symbol)
 		/// </summary>
 		/// <param name="symbol">The current symbol</param>
 		/// <param name="typeName">A potential base class.</param>
-		public static bool Is(this INamedTypeSymbol? symbol, string typeName)
+		public static bool InheritsFrom(this INamedTypeSymbol? symbol, string typeName)
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/SymbolExtensions.cs:94]: I have not problem with this change, as it makes it more readable, but it'll be better if you make it a different PR.

Comment (Youssef1313) [reply src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/SymbolExtensions.cs:94]: @jeromelaban Will do. I'll revert the change and have it in a new PR. Going to squash that along with proper testing. Would be helpful if you or someone can guide me to the proper way to testing this.

Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/SymbolExtensions.cs:94]: I don't think you'll be able to make a unit test out of this because of platform dependencies, though you can probably make one [in here](https://github.com/unoplatform/uno/tree/master/src/SourceGenerators/XamlGenerationTests), as it is built to validate that the generated code can compile, or make add it to the [runtime tests](https://github.com/unoplatform/uno/tree/master/src/Uno.UI.RuntimeTests) assembly, even if it's not really testing anything at runtime.

Comment (Youssef1313) [reply src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/SymbolExtensions.cs:94]: @jeromelaban The first link is for `XamlGenerationTests`, does it support native ctor generation testing too?

Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/SymbolExtensions.cs:94]: It runs all the generators, so yes, it will validate this as well. But after reflection, it's best if you add it to the runtime tests.

**Comment by jeromelaban** [comment]
The build is failing for this reason:

```
       (CoreCompile target) -> 
         obj/Release/monoandroid11.0/g/NativeCtorsGenerator/Uno_UI_RuntimeTests_Tests_NativeCtorGeneratorTests_GameBoardCanvas_414dd57b398860c99739545435d09fad.g.cs(28,109): error CS1503: Argument 1: cannot convert from 'System.IntPtr' to 'Android.Content.Context' [/Users/runner/work/1/s/src/Uno.UI.RuntimeTests/Uno.UI.RuntimeTests.csproj]
         obj/Release/monoandroid11.0/g/NativeCtorsGenerator/Uno_UI_RuntimeTests_Tests_NativeCtorGeneratorTests_GameBoardCanvas_414dd57b398860c99739545435d09fad.g.cs(28,124): error CS1503: Argument 2: cannot convert from 'Android.Runtime.JniHandleOwnership' to 'Android.Util.IAttributeSet' [/Users/runner/work/1/s/src/Uno.UI.RuntimeTests/Uno.UI.RuntimeTests.csproj]
```

Looks like the test failed for a good reason :)

File: src/Uno.UI.RuntimeTests/Tests/NativeCtorGeneratorTests/GameBoardCanvas.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿#if MONOANDROID
```
Comment (jeromelaban) [inline src/Uno.UI.RuntimeTests/Tests/NativeCtorGeneratorTests/GameBoardCanvas.cs:1]: You can use 
```suggestion
﻿#if __ANDROID__
```
or suffix the file with `.android.cs`

**Comment by Youssef1313** [comment]
> /Users/runner/work/1/s/src/Uno.UI.RuntimeTests/Tests/NativeCtorGeneratorTests/GameBoardCanvas.cs(6,41): error CS0246: The type or namespace name 'SkiaSharp' could not be found (are you missing a using directive or an assembly reference?) [/Users/runner/work/1/s/src/Uno.UI.RuntimeTests/Uno.UI.RuntimeTests.net6.csproj]

I'm not sure why it's not found. The package reference exists.

File: src/Uno.UI.RuntimeTests/Tests/NativeCtorGeneratorTests/GameBoardCanvas.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#if __ANDROID__
+using Android.Content;
+
+namespace Uno.UI.RuntimeTests.Tests.NativeCtorGeneratorTests
+{
+	public partial class GameBoardCanvas : SkiaSharp.Views.Android.SKGLTextureView
```
Comment (jeromelaban) [inline src/Uno.UI.RuntimeTests/Tests/NativeCtorGeneratorTests/GameBoardCanvas.cs:6]: Let's take this a different way. Why do you need this specific type ? `SKGLTextureView` ? Can't you use one that comes with the android API?

**Comment by jeromelaban** [comment]
Looks like there's a missing partial modifier :)

**Comment by jeromelaban** [comment]
> @jeromelaban Is this ready to merge?

Not until the build passes :)

**Comment by Youssef1313** [comment]
@jeromelaban I assume the following failure:

    56>MSBUILD : error : Generation failed: System.IO.FileLoadException: Could not load file or assembly 'Microsoft.CodeAnalysis.Workspaces, Version=3.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. Access is denied. [C:\a\1\s\src\AddIns\Uno.UI.Lottie\Uno.UI.Lottie.Skia.csproj]

is a flakiness. Can you rerun the failing check? (*currently* it's only this one failing)

**Review by jeromelaban — APPROVED** [review]
Very nice! Thanks for your contribution!
